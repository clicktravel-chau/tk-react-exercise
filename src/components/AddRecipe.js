import React, { useState } from 'react';
import RecipeService from '../services/RecipeService';

const AddRecipe = () => {
  const initialRecipeState = {
    id: null,
    name: '',
    description: '',
    ingredients: '',
  };

  const [recipe, setRecipe] = useState(initialRecipeState);
  const [submitted, setSubmitted] = useState(false);

  const handleInputChange = event => {
    const { name, value } = event.target;
    setRecipe({ ...recipe, [name]: value });
  };

  const saveRecipe = () => {
    const ingredients = recipe.ingredients
      .split(',')
      .map(ingredient => ({ name: ingredient.trim() }));

    const data = {
      name: recipe.name,
      description: recipe.description,
      ingredients: ingredients,
    };

    RecipeService.create(data)
      .then(response => {
        setRecipe({
          id: response.data.id,
          name: response.data.name,
          description: response.data.description,
          ingredients: response.data.ingredients,
        });

        setSubmitted(true);
        console.log(response.data);
      })
      .catch(e => {
        console.log(e);
      });
  };

  const newRecipe = () => {
    setRecipe(initialRecipeState);
    setSubmitted(false);
  };

  return (
    <div data-testid="add-recipe-root" className="submit-form">
      {submitted ? (
        <div>
          <h4 data-testid="success-header">You submitted successfully!</h4>
          <button className="btn btn-success" onClick={newRecipe}>
            Add another
          </button>
        </div>
      ) : (
        <div>
          <div className="form-group">
            <label htmlFor="name">Name</label>
            <input
              type="text"
              className="form-control"
              id="name"
              required
              value={recipe.name}
              onChange={handleInputChange}
              name="name"
            />
          </div>

          <div className="form-group">
            <label htmlFor="description">Description</label>
            <input
              type="text"
              className="form-control"
              id="description"
              required
              value={recipe.description}
              onChange={handleInputChange}
              name="description"
            />
          </div>

          <div className="form-group">
            <label htmlFor="ingredients">Ingredients (separate them with a comma)</label>
            <input
              type="text"
              className="form-control"
              id="ingredients"
              required
              value={recipe.ingredients}
              onChange={handleInputChange}
              name="ingredients"
            />
          </div>

          <button onClick={saveRecipe} className="btn btn-success">
            Submit
          </button>
        </div>
      )}
    </div>
  );
};

export default AddRecipe;
