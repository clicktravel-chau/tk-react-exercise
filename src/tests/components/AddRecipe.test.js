import React from 'react';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import axiosMock from 'axios';
import AddRecipe from '../../components/AddRecipe';

describe('AddRecipe', () => {
  afterEach(cleanup);

  it('renders', () => {
    const { asFragment } = render(<AddRecipe />);
    expect(asFragment()).toMatchSnapshot();
  })

  describe('has form for adding a recipe', () => {

    it('should be a class for a submittable form containing name, description and ingredients',
      () => {
        const {
          getByTestId,
          getByLabelText,
        } = render(<AddRecipe />);

        expect(getByTestId('add-recipe-root')).toHaveClass('submit-form');
        expect(getByLabelText('Name')).toBeVisible()
        expect(getByLabelText('Description')).toBeVisible();
        expect(getByLabelText('Ingredients (separate them with a comma)')).toBeVisible();
      });

    it('should submit successfully', async () => {
      // Given
      axiosMock.post.mockResolvedValue({
        data: {
          id: 1,
          name: 'Pizza',
          description: 'Bake in the oven',
          ingredients: [{ name: 'cheese' }],
        },
      });

      // When
      const {
        getByTestId,
        getByText,
      } = render(<AddRecipe />);
      fireEvent.click(getByText('Submit'));

      // Then
      await waitFor(() => expect(getByTestId('success-header')).toHaveTextContent(
        'You submitted successfully!'));
      expect(axiosMock.post).toHaveBeenCalledTimes(1);
    });
  });
});
