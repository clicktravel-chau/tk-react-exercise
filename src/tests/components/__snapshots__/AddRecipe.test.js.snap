// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipe renders 1`] = `
<DocumentFragment>
  <div
    class="submit-form"
    data-testid="add-recipe-root"
  >
    <div>
      <div
        class="form-group"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          class="form-control"
          id="name"
          name="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="description"
        >
          Description
        </label>
        <input
          class="form-control"
          id="description"
          name="description"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="ingredients"
        >
          Ingredients (separate them with a comma)
        </label>
        <input
          class="form-control"
          id="ingredients"
          name="ingredients"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-success"
      >
        Submit
      </button>
    </div>
  </div>
</DocumentFragment>
`;
